xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/js/xwiki/xwiki.js

```diff
@@ -943,235 +943,206 @@ function checkAdvancedContent(message) {
-                    'target':document,
-                    'keycode':false
-            }
-            if(!opt) opt = default_options;
-            else {
-                for(var dfo in default_options) {
-                    if(typeof opt[dfo] == 'undefined') opt[dfo] = default_options[dfo];
-                }
-            }
+shortcut = new Object({
 
-            var ele = opt.target
-            if(typeof opt.target == 'string') ele = document.getElementById(opt.target);
-            var ths = this;
-            shortcut_combination = shortcut_combination.toLowerCase();
-
-            //The function to be called at keypress
-            var func = function(e) {
-                e = e || window.event;
-
-                if(opt['disable_in_input']) { //Don't enable shortcut keys in Input, Textarea fields
-                    var element;
-                    if(e.target) element=e.target;
-                    else if(e.srcElement) element=e.srcElement;
-                    if(element.nodeType==3) element=element.parentNode;
-                    if(element.tagName == 'INPUT' || element.tagName == 'TEXTAREA' || element.tagName == 'SELECT') {
-                        return;
-                    }
-                }
+    /**
+     * @returns {Array} of registered shortcuts
+     */
+    all_shortcuts: function() {
+        var shortcuts = [];
+
+        Object.values(this._listeners).forEach(function(group, index) {
+            Object.values(group).forEach(function(listener, index2) {
+                shortcuts = shortcuts.concat(listener.get_registered_combos());
+            })
+        });
+
+        return shortcuts;
+    },
+
+    /**
+     * Type of shortcut that can be registered.
+     */
+    type: new Object({
+        SIMPLE: 'simple',
+        COUNTING: 'counting',
+        SEQUENCE: 'sequence'
+    }),
+
+    /**
+     * Add a new shortcut.
+     *
+     * The opt parameter should be a map of optional parameters used while registering the shortcut.
+     * <ul>
+     *     <li>target: A DOM element in which the shortcut should be listened (defaults to the whole document)</li>
+     *     <li>disable_in_input: If true, the shortcut will not be listened when an input
+     *     or textarea field is selected (default: false)</li>
+     *     <li>type: The type (shortcut.type) that should be used (default: SIMPLE)</li>
+     * </ul>
+     *
+     * @param shortcut_combination {string} the shortcut that should trigger the callback
+     * @param callback the function triggered by the shortcut
+     * @param opt optional associative array defining parameters for registering the shortcut
+     */
+    add: function(shortcut_combination, callback, opt) {
+        // Require Keypress JS to be fully loaded before registering the shortcut.
+        require(["$services.webjars.url('org.webjars.bower:Keypress', 'keypress.js')", 'jquery'], function(keypress) {
+
+            // If no options are defined, create a blank array
```

Comment:
Isn't it semantically more correct to use a map `{}` instead of an array `[]`?

Reply:
Yep, will fix that, thanks :)